src/NServiceBus.Core/Sagas/SagaMetadata.cs

```diff
@@ -110,5 +119,273 @@ public bool TryGetFinder(string messageType, out SagaFinderDefinition finderDefi
+        public static SagaMetadata Create(Type sagaType)
+        {
+            return Create(sagaType, new List<Type>(), new Conventions());
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SagaMetadata" /> from a specific Saga type.
+        /// </summary>
+        /// <param name="sagaType">A type representing a Saga. Must be a non-generic type inheriting from <see cref="NServiceBus.Saga.Saga" />.</param>
+        /// <param name="availableTypes">Additional available types, used to locate saga finders and other related classes.</param>
+        /// <param name="conventions">Custom conventions to use while scanning types.</param>
+        /// <returns>An instance of <see cref="SagaMetadata" /> describing the Saga.</returns>
+        public static SagaMetadata Create(Type sagaType, IEnumerable<Type> availableTypes, Conventions conventions)
+        {
+            if (!IsSagaType(sagaType))
+            {
+                throw new Exception(sagaType.FullName + " is not a saga");
+            }
+
+            var genericArguments = GetBaseSagaType(sagaType).GetGenericArguments();
+            if (genericArguments.Length != 1)
+            {
+                throw new Exception(string.Format("'{0}' saga type does not implement Saga<T>", sagaType));
+            }
+
+            var saga = (Saga)FormatterServices.GetUninitializedObject(sagaType);
+            var mapper = new SagaMapper();
+            saga.ConfigureHowToFindSaga(mapper);
+
+            var sagaEntityType = genericArguments.Single();
+
+            ApplyScannedFinders(mapper, sagaEntityType, availableTypes, conventions);
+
+            var correlationProperties = new List<CorrelationProperty>();
+            var finders = new List<SagaFinderDefinition>();
+
+            foreach (var mapping in mapper.Mappings)
+            {
+                if (!mapping.HasCustomFinderMap)
+                {
+                    correlationProperties.Add(new CorrelationProperty(mapping.SagaPropName));
+                }
+
+                SetFinderForMessage(mapping, sagaEntityType, finders);
+            }
+
+            var associatedMessages = GetAssociatedMessages(sagaType)
+                .ToList();
+
+            return new SagaMetadata(sagaType.FullName, sagaType, sagaEntityType.FullName, sagaEntityType, correlationProperties, associatedMessages, finders);
+        }
+
+        static void ApplyScannedFinders(SagaMapper mapper, Type sagaEntityType, IEnumerable<Type> availableTypes, Conventions conventions)
+        {
+            var actualFinders = availableTypes.Where(t => typeof(IFinder).IsAssignableFrom(t) && t.IsClass)
+                .ToList();
+
+            foreach (var finderType in actualFinders)
+            {
+                foreach (var interfaceType in finderType.GetInterfaces())
+                {
+                    var args = interfaceType.GetGenericArguments();
```

Comment:
I'd rather have the body of this loop extracted so that I can easier follow the idea of this method.